### PR TITLE
lispPackagesLite: pure Nix: Common Lisp modules without QuickLisp

### PR DESCRIPTION
**Pull request by hraban**
Full docs in the [README](https://github.com/hraban/nixpkgs/blob/feat/lisp-packages-lite/pkgs/development/lisp-modules-lite#readme).

###### Description of changes

This is a POC of a pure Nix implementation of Common Lisp modules, without QuickLisp or pre-compilation or anything of the sort.

It is definitely WIP but I'm sharing it in case others find it interesting. Comments, thoughts and questions are more than welcome.

This is the real solution to https://github.com/NixOS/nixpkgs/pull/196818 (or the bug it tried to fix).

The README has the summary of what this is, and the source code has a lot more info.

Please forgive the TODOs and any potential obsolete or unprofessional comments. As I said: it's very WIP.

The merge base is master because that makes it easier to share this work for now, and to work on it, locally.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by 7c6f434c** [comment]
So the dependencies need to be entered manually, like before the Quicklisp-scraping helpers?

**Comment by hraban** [comment]
Yeah, this is a POC to evaluate what it's like to live without QuickLisp.

I idn't put this in the rationale but maybe I should: I did some digging into quicklisp before going down this road, and realized that QuickLisp:

- doesn't offer SSL
- doesn't offer SHA
- Offers MD5 but doesn't actually ever check that, afai can tell¹
- All depends on 1 man (pretty crucial bus factor)
- Is upgraded at Zach's behest, not when the community thinks it's best
- Adding new packages: again, all at Zach's discretion

It led me down a rabbit hole where I realized that QuickLisp is a solution to a world of Common Lisp without Nix. Nix itself is a better quicklisp, and it's worth seeing what it looks like to embrace that.

While Googling this I came across this Hacker News comment that put me on to this thinking: https://news.ycombinator.com/item?id=13097333

Note from that thread:

>  I hope to improve the documentation and security of Quicklisp in the next few months. It will depend on funding.
- Zach, Dec 2016

Almost 6 years later: still HTTP, still MD5, and that's assuming the hashes are even checked.

The main benefit of QuickLisp is "just list your dependencies and you are good to go." But Nix offers the same.

Not to cast any shade on Zach! I met him personally and he was the loveliest person you could imagine him to be. You pretty much have to be, to volunteer this much time to the community. I'm very grateful, this isn't a protest against QuickLisp. Just an acknowledgement that Nix is a solution to the same problem, worth exploring.

---

¹  Please don't pin me on this, I would want to confirm with Zach before committing libel :). I looked around the QL codebase and didn't find any MD5 verification. The clearest hint I have that no hashes are checked is that afaik, he'd need to include an MD5 hash implementation in quicklisp for bootstrapping, and I haven't seen one.

File: pkgs/development/lisp-modules-lite/default.nix [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+        lispDependencies = map (dep: dep.overrideSystemsMap sourceToSystems) allMyDeps;
+        lisp = sbcl;
+        lispSystems = b.attrNames sourceToSystems.${mySrcDrv}.systems;
+      });
+
+  # If a single src derivation specifies multiple lisp systems, you can use this
+  # helper to define them.
+  lispMultiDerivation = args: a.mapAttrs (name: system:
+    let
+      namearg = a.optionalAttrs (! a.hasAttrByPath ["lispSystems"] system) { lispSystem = name; };
```
Comment (hraban) [inline pkgs/development/lisp-modules-lite/default.nix:263]: is there no `builtins.elem` equivalent for attrsets?

**Comment by 7c6f434c** [comment]
>- Is upgraded at Zach's behest, not when the community thinks it's best

With consistent testing, though…

>- Adding new packages: again, all at Zach's discretion

(Is this discretion ever actually exercised? As opposed to adding the package to the addition queue as soon as the issue fulfills the basic completeness criteria)

>Almost 6 years later: still HTTP, still MD5, and that's assuming the hashes are even checked.

SHA1 is actually provided. Checked, now that's a better question. (I think neither MD5 nor SHA1 is checked, yeah)

>The main benefit of QuickLisp is "just list your dependencies and you are good to go." But Nix offers the same.

Nope. As show the scars in _our_ infrastructure, Quicklisp works pretty well even if the deps are mislisted. Well, you need to specify things in ASDF, but this can rely on arbitrary additional methods to ASDF generic functions…

>Not to cast any shade on Zach! I met him personally and he was the loveliest person you could imagine him to be. You pretty much have to be, to volunteer this much time to the community. I'm very grateful, this isn't a protest against QuickLisp. Just an acknowledgement that Nix is a solution to the same problem, worth exploring.

By just entering more information by hand than needs to be entered by hand into quicklisp project list?

**Comment by hraban** [comment]
Yes, I think we're in agreement on most of those points. When I say "at Zach's discretion" I mean that factually, not disparagingly. He's clearly been a good steward of the project.

The only point I didn't understand was:

> Nope. As show the scars in _our_ infrastructure, Quicklisp works pretty well even if the deps are mislisted. Well, you need to specify things in ASDF, but this can rely on arbitrary additional methods to ASDF generic functions…

What do you mean about deps being mislisted?

As for the SHA you are right, it's SHA1, though unlike the MD5 it's not of the package itself just the contents. I couldn't figure out what that actually meant--in fact, trying to discover what that SHA1 represented was what led me down this rabbit hole in the first place.

As for this:

> By just entering more information by hand than needs to be entered by hand into quicklisp project list?

Yes, I think we mean the same thing, yes. To some people, this is a feature, to others it's an anti-feature. To be perfectly frank, listing repository and version information was by far the least time consuming part of this. And given the "restrained" pace at which most CL packages are updated (👍), I'd be pretty happy to even just use this alone for my own projects.

For anything mission critical this feels like a more reliable approach than QL (again see the HN thread on this)

Let me emphasize this is just a POC, an exploration. :)

Lastly:

> With consistent testing, though…

Thank you I totally forgot to add that to the TODOs: next on the agenda is exploring registering packages' test suites and running them from within Nix.

(Btw: technically, this approach does two things: rewriting of the actual lisp module implementation, and discarding QL entirely. It seemed like a worthwile avenue to explore, but it is theoretically entirely compatible with fetching sources from QL, as the others do. Which versions are used is orthogonal to the actual implementation of the module system, and I can make "package version overriding" work well enough, it should even be doable entirely from "user land". In fact, that'd be a nice litmus test!)

**Comment by hraban** [comment]
To be abundantly clear: I'm working on this *for me*. I'm sharing it with the world for the shake of sharing. The easiest way to do that is a draft PR, but I see it as entirely reasonable for this to eventually live separately and never be merged.

This is not "the other implementations are wrong" or "QuickLisp is stupid".

**Comment by 7c6f434c** [comment]
>> Nope. As show the scars in _our_ infrastructure, Quicklisp works pretty well even if the deps are mislisted. Well, you need to specify things in ASDF, but this can rely on arbitrary additional methods to ASDF generic functions…
>
>What do you mean about deps being mislisted?

Quicklisp lists dependencies of systems in its own system list file. These are autogenerate from the package content, but also these are just hints.

If the deps end up wrong… nothing noticeable happens, because there are hooks inserted into ASDF to fallback to just download the necessary tarball if it wasn't fetch as a listed dependency. So converting the dep list by looking at Quicklisp usually works, but it can fail while Quicklisp doesn't.

>As for this:
>
>> By just entering more information by hand than needs to be entered by hand into quicklisp project list?
>
>Yes, I think we mean the same thing, yes. To some people, this is a feature, to others it's an anti-feature. 

I mean, for various obvious reasons we already hae some lisp packages defined manually in each version of quicklisp-based tooling…

>Let me emphasize this is just a POC, an exploration. :)

Well, the question is a PoC _of what_… 

> To be abundantly clear: I'm working on this *for me*.

… and given that in any case Quicklisp fetcher generates code for Quicklisp-agnostic lower layer, whether it is useful to use it as a replacement lower layer or not.

**Comment by hraban** [comment]
There's much to digest in this reply so I will ponder a while and get back to you after more experimentation.

Meanwhile:

> Well, the question is a PoC _of what_…

The core of this implementation is the automatic deduplication of lisp dependencies when two separate systems are supplied by the same source derivation. This is a way to solve "the cl-async issue" which is present in lisp-modules-new (because it includes the same source multiple times for different systems).

The systems.txt file makes no distinction between "exported" and "private" systems, which is a problem when combined with an immutable /nix/store/.

**Comment by 7c6f434c** [comment]
> The core of this implementation is the automatic deduplication of lisp dependencies when two separate systems are supplied by the same source derivation.

Does this mean that pre-compilation work is never shared between different «top-level» Common Lisp «targets» (like sets of packages for a shell or something)?

**Comment by hraban** [comment]
> > The core of this implementation is the automatic deduplication of lisp dependencies when two separate systems are supplied by the same source derivation.
> 
> Does this mean that pre-compilation work is never shared between different «top-level» Common Lisp «targets» (like sets of packages for a shell or something)?

No, that work is in fact shared. It means that if your derivation supplies e.g. optima and optima-ppcre, then if a child derivation includes both separately, the system will find both of them, unify them into a single new derivation (systems-optima_optima-ppcre) with all unified dependencies and compile that. If an entirely different compile relies on both of those as well, it will reuse that.

The source code has a lot of comments, and hopefully it should explain the details.

Btw I'm working on a v2 which offers the same functionality but with a slightly different implementation.

**Comment by hraban** [comment]
Update: Over the last few weeks I have added some considerable features:

- tests
- more packages
- overridePackage support
- More documentation, including a demo of flake support

I have also created a script for updating sources ("https://github.com/hraban/nix-update-sources"). It's only a POC, but it demonstrates that this is possible. It also dogfoods the project.

**Comment by hraban** [comment]
> Quicklisp lists dependencies of systems in its own system list file. These are autogenerate from the package content, but also these are just hints. If the deps end up wrong… nothing noticeable happens, because there are hooks inserted into ASDF to fallback to just download the necessary tarball if it wasn't fetch as a listed dependency. So converting the dep list by looking at Quicklisp usually works, but it can fail while Quicklisp doesn't.

I now understand what you mean, and I personally consider this an anti feature *if I'm using Nix*. If I'm just using QL, that's awesome. Just Do The Right Thing. But in Nix, and *especially* with flakes (and its flake.lock), I am taking explicit control of dependency management.

Now that I have more experience with adding a large number of packages, I must say it's not that much work. Some are tricky, most are straight-forward, once it's done it's done. Barring updates of course.

> I mean, for various obvious reasons we already hae some lisp packages defined manually in each version of quicklisp-based tooling…
This sentence makes sense to me but I'm confused about the specific thread it appears in.

Maybe it will help if I quote the essence of this project here from the readme:

*This is an experiment to discover what a Lisp-in-Nix system would look like, if QuickLisp didn’t exist. **I started with a `stdenv.mkDerivation` and worked my up from there.***

I'm leaning hard into Nix & ASDFv3, and I noticed that Nix & ASDFv3 have some tricky incompatibilities which require very very careful management. Deduplication of dependencies is essential, including the same source multiple times is asking for trouble, and mangling .asd files (what I originally tried in NixOS/nixpkgs#196818) or inspecting them in any way, or trying to be smart about lisp code in any automated way, is fraught with peril and edge cases.

QuickLisp's effort in curating packages and finding updates is useful and can be used in a manual process to update the package tree in this project. But I have come to the conclusion it cannot be an entirely automated process, because QuickLisp (using ASDFv3) is happy to write cached .fasl files at will, in circular dependencies (of source, not of systems, this gets tricky but trust me or read this module's Nix source :)). That approach is fundamentally at odds with Nix, which has two constraints:

- don't write in /nix/store, iow don't write to dependencies
- don't write in your home dir

This is very hard to reconcile with ASDF, and basically impossible to reconcile with QuickLisp, because quicklisp doesn't have that kind of metadata, because quicklisp doesn't care.

For more details, please feel free to ask questions, or have a look through the source of this project, it has a lot of comments from my painful hours of debugging 🥲

**Comment by 7c6f434c** [comment]
>> Quicklisp lists dependencies of systems in its own system list file. These are autogenerate from the package content, but also these are just hints. If the deps end up wrong… nothing noticeable happens, because there are hooks inserted into ASDF to fallback to just download the necessary tarball if it wasn't fetch as a listed dependency. So converting the dep list by looking at Quicklisp usually works, but it can fail while Quicklisp doesn't.
>
>I now understand what you mean, and I personally consider this an anti feature *if I'm using Nix*. If I'm just using QL, that's awesome. Just Do The Right Thing. But in Nix, and *especially* with flakes (and its flake.lock), I am taking explicit control of dependency management.

Well, the entire cursed magic in quicklispToNix is to let QL do its work and then extract deps in a Nix-friendly form.

>Now that I have more experience with adding a large number of packages, I must say it's not that much work. Some are tricky, most are straight-forward, once it's done it's done. Barring updates of course.

Work per package is much more than if you can get _most_ through a consistent source.

>*This is an experiment to discover what a Lisp-in-Nix system would look like, if QuickLisp didn’t exist. **I started with a `stdenv.mkDerivation` and worked my up from there.***

This looks like a simple misunderstanding of all the other infrastructure versions' relation to Quicklisp.

Like the real thing is your approach to the question of which-asd-file via deduplication. What you declare a difference is just a lack of a useful feature that can be added on top without changing your design.

>This is very hard to reconcile with ASDF, and basically impossible to reconcile with QuickLisp, because quicklisp doesn't have that kind of metadata, because quicklisp doesn't care.

Well we definitely don't want to run Quicklisp within Nix builds; we do need overrides after export from Quicklisp

**Comment by hraban** [comment]
> This looks like a simple misunderstanding of all the other infrastructure versions' relation to Quicklisp.

Do you mean lisp-packages-new loads every QL system independently, discards the QL supplied dependency info, instead lists the systems actually loaded into ASDF, and dumps that as the *real* dependency info? If that's the case then I agree with you: I misunderstood. That would certainly be an interesting and valuable approach, and it would be a cool challenge to solve this for different lisps and different target systems (packages have different dependencies depending on host OS and arch).

If, rather, lisp-packages-new just accepts the dependency info from QL, then... I don't understand what you mean here. Where's the misunderstanding?


> Like the real thing is your approach to the question of which-asd-file via deduplication. What you declare a difference is just a lack of a useful feature that can be added on top without changing your design.

I mean... yeah, when I read this sentence it almost sounds like we're disagreeing, but I completely agree with this and it was my point the whole time. Sorry for expressing myself poorly?

**Comment by 7c6f434c** [comment]
>> This looks like a simple misunderstanding of all the other infrastructure versions' relation to Quicklisp.
>
>Do you mean lisp-packages-new loads every QL system independently, discards the QL supplied dependency info, instead lists the systems actually loaded into ASDF, and dumps that as the *real* dependency info? If that's the case then I agree with you: I misunderstood. That would certainly be an interesting and valuable approach, and it would be a cool challenge to solve this for different lisps and different target systems (packages have different dependencies depending on host OS and arch).
>
>If, rather, lisp-packages-new just accepts the dependency info from QL, then... I don't understand what you mean here. Where's the misunderstanding?

Old quicklispPackages do load-then-analyse with Quicklisp to generate the dependency list. -new does trust Quicklisp more — but it still just uses it as a first iteration of dependency data, in update-time, not in buid-time.

>> Like the real thing is your approach to the question of which-asd-file via deduplication. What you declare a difference is just a lack of a useful feature that can be added on top without changing your design.
>
>I mean... yeah, when I read this sentence it almost sounds like we're disagreeing, but I completely agree with this and it was my point the whole time. Sorry for expressing myself poorly?

I mean, good if we are agreing, then it just mean you should not stress Quicklisp part more than «not yet implemented or specifically planned».

(And so we are back to my wondering if it is in a good shape for drop-in replacement of the actual build-time for lispPackages, with dependency data still mostly managed via Quicklisp dumpers we already have, with some manual overrides and some manually added packages)

**Comment by hraban** [comment]
> Old quicklispPackages do load-then-analyse with Quicklisp to generate the dependency list. -new does trust Quicklisp more — but it still just uses it as a first iteration of dependency data, in update-time, not in buid-time.

Certainly interesting, I confess I know less about the original lispPackages than the -new one. I don't know exactly what you mean about the "first iteration" part, from what I gathered it was fully using QL to populate the sqlite, and extracting that info the generate the .nix files, but I trust you when you say there's more stuff going on.

> I mean, good if we are agreing, then it just mean you should not stress Quicklisp part more than «not yet implemented or specifically planned»

Just for my reference: did you read the linked README? I don't mean to give you homework, it's just useful context for me to know what to mention, and what to leave out.

> (And so we are back to my wondering if it is in a good shape for drop-in replacement of the actual build-time for lispPackages, with dependency data still mostly managed via Quicklisp dumpers we already have, with some manual overrides and some manually added packages)

I hope I never gave the impression that this was intended as a drop in replacement for any of the two current implementations! :) In its current state this is a POC, an experiment. Even in its most theoretical best case best effort extreme, this is an exercise in different trade-offs. It's all about trade-offs. This would be a parallel or separate thing, that people are free to play with and choose over QL if they prefer.

This is a draft PR on purpose. In fact, one of my TODOs is "figure out if this belongs in nixpkgs at all", and that's not even considering "if so, figure out if other people agree" which is just as important if not more.

I would like to take this opportunity to invite you to have a real substantive look at the code and the examples. We can keep going back and forth about the way I've proposed it, but we'll keep arguing between your impression of what it is, and my inability to articulate it, neither of which might be very helpful. If you have the time: dive in, see what you think. What do you think of the deduplication mechanism? If you would use QL to update the fixtures: how would you do it? What do you think of the makeScope? Etc.

We're getting very focused on "how I talk about it" rather than "what it actually is". If you have the time, have a look, let me know what you find. 😀

**Comment by 7c6f434c** [comment]
README does not really add much clarity for me.

Basically: if it is a draft but PR to Nixpkgs, I want to plan for it either being used as a replacement lower level, or existing in parallel, and then I want to understand when you want which version. I don't think drop-in (or minimal-wrapper) lower level has that many requirements per se, although maybe more complicated evaluation is unfortunate.

It is definitely the least «lite» version, though, as it does interesting but complicated work in Nix (… to read the Nix code, I prefer to be sure it is reviewed against what exact goal, of course, thus asking which goals are the real load-bearing ones; looks like it ensures «foo-b & foo-c» is taken in preference to «foo-c» in a weird dependency chain just because of registry sorting, is this right?)

Or is it not supposed to be merged into Nixpkgs (e.g. because of a different license for Nix code)? Seems to plan otherwise, but again, I might be easily confused by this README.

**Pull request closed without merging**